### PR TITLE
feat: add build_model function

### PR DESCRIPTION
**Pull request by LMMilliken**
This PR adds a `build_model` using features added in [pr 336](https://github.com/jina-ai/finetuner-core/pull/336).

---

- [ ] This PR references an open issue
- [x] I have added a line about this change to CHANGELOG

### CODE REVIEW

**Review by bwanglzu — COMMENTED** [review]
minor comments

File: finetuner/__init__.py [hunk 1]
```diff
@@ -308,6 +308,63 @@ def get_token() -> str:
...
+    :param device: Whether to use the CPU, if set to `cuda`, a Nvidia GPU will be used.
+        otherwise use `cpu` to run a cpu job.
+    :param logging_level: The executor logging level. See
+        https://docs.python.org/3/library/logging.html#logging-levels for available
+        options.
+    :param is_onnx: The model output format, either `onnx` or `pt`.
+    :param model_options: A dictionary of model specific options
+    :return: A built pretrained model
+    """
+    from commons.models.inference import (
```
Comment (gmastrapas) [inline finetuner/__init__.py:341]: from finetuner.commons.models ...

File: finetuner/__init__.py [hunk 2]
```diff
@@ -308,6 +308,63 @@ def get_token() -> str:
...
+        ONNXRuntimeInferenceEngine,
+        TorchInferenceEngine,
+    )
+    from commons.runner.model import RunnerModel
```
Comment (gmastrapas) [inline finetuner/__init__.py:345]: from finetuner.commons.runner ...

File: finetuner/__init__.py [hunk 3]
```diff
@@ -308,6 +308,63 @@ def get_token() -> str:
...
+    logging_level: str = 'DEBUG',
+    is_onnx: bool = False,
+    model_options: Dict[str, Any] = {},
+):
+    """
+    Builds a pre-trained model from a given descriptor.
+
+    :param name: Refers to a pre-trained model, see
+        https://finetuner.jina.ai/walkthrough/choose-backbone/ for a list of all
+        supported models
```
Comment (gmastrapas) [inline finetuner/__init__.py:325]: period in the end

File: finetuner/__init__.py [hunk 4]
```diff
@@ -308,6 +308,63 @@ def get_token() -> str:
...
+    :param select_model: Finetuner run artifacts might contain multiple models. In
+        such cases you can select which model to deploy using this argument. For CLIP
+        fine-tuning, you can choose either `clip-vision` or `clip-text`.
+    :param device: Whether to use the CPU, if set to `cuda`, a Nvidia GPU will be used.
+        otherwise use `cpu` to run a cpu job.
+    :param logging_level: The executor logging level. See
+        https://docs.python.org/3/library/logging.html#logging-levels for available
+        options.
+    :param is_onnx: The model output format, either `onnx` or `pt`.
+    :param model_options: A dictionary of model specific options
```
Comment (gmastrapas) [inline finetuner/__init__.py:338]: period in the end. `model_options` should be the second argument

**Review by gmastrapas — CHANGES_REQUESTED** [review]
1. Also update changelog
2. Include a unit test for this method

File: finetuner/__init__.py (hunk 2 above)
Comment (LMMilliken) [reply finetuner/__init__.py:345]: Imports in other methods in this file import from commons like this, are those also incorrect?

File: finetuner/__init__.py [hunk 5]
```diff
@@ -308,6 +308,64 @@ def get_token() -> str:
...
+    select_model: Optional[str] = None,
+    device: str = 'cpu',
+    logging_level: str = 'DEBUG',
+    is_onnx: bool = False,
+):
+    """
+    Builds a pre-trained model from a given descriptor.
+
+    :param name: Refers to a pre-trained model, see
+        https://finetuner.jina.ai/walkthrough/choose-backbone/ for a list of all
```
Comment (guenthermi) [inline finetuner/__init__.py:324]: I think it is better to refer to the `finetuner.describe_models()` function or mention both. In the documentation we might not display the whole table anymore at some point

File: tests/unit/test___init__.py [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+                select_model=select_model,
+                is_onnx=is_onnx,
+            )
+    else:
+        model = finetuner.build_model(
+            name=descriptor, select_model=select_model, is_onnx=is_onnx
+        )
+    # encode and check the embeddings
+    model = finetuner.build_model(
+        name='bert-base-cased',
```
Comment (guenthermi) [inline tests/unit/test___init__.py:34]: is this correct to call it a second time?

Comment (LMMilliken) [reply tests/unit/test___init__.py]: no it is not, fixed now

File: CHANGELOG.md [hunk 7]
```diff
@@ -10,6 +10,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ### Added
 
+- Add `build_model` function to create zero-shot models ([#584](https://github.com/jina-ai/finetuner/pull/584))
```
Comment (gmastrapas) [inline CHANGELOG.md:13]: dot in the end

File: finetuner/__init__.py [hunk 8]
```diff
@@ -308,6 +308,69 @@ def get_token() -> str:
     return ft.get_token()
 
 
+def build_model(
+    name: str,
+    model_options: Dict[str, Any] = {},
+    batch_size: int = 32,
+    select_model: Optional[str] = None,
+    device: Optional[str] = None,
+    logging_level: str = 'DEBUG',
```
Comment (gmastrapas) [inline finetuner/__init__.py:317]: lets remove this argument, it seems irrelevant. Also remove it from docstring

File: finetuner/__init__.py [hunk 9]
```diff
@@ -308,6 +308,69 @@ def get_token() -> str:
     return ft.get_token()
 
 
+def build_model(
+    name: str,
+    model_options: Dict[str, Any] = {},
```
Comment (gmastrapas) [inline finetuner/__init__.py:313]: having a deafult empty dict is dangerous. You need to set it as Optional and to None by default. Then inside the function you can do `model_options or {}`

**Review by bwanglzu — APPROVED** [review]
LGTM!

File: finetuner/__init__.py [hunk 10]
```diff
@@ -308,6 +308,69 @@ def get_token() -> str:
...
+    from _finetuner.runner.model import RunnerModel
+
+    if not device:
+        device = 'cuda' if torch.cuda.is_available() else 'cpu'
+
+    stub = model_stub.get_stub(
+        name, select_model=select_model, model_options=model_options
+    )
+
+    runner = RunnerModel(stub=stub)
```
Comment (gmastrapas) [inline finetuner/__init__.py:357]: thats not a runner, its a model

File: setup.py [hunk 11]
```diff
@@ -27,6 +27,7 @@
         zip_safe=False,
         setup_requires=['setuptools>=18.0', 'wheel'],
         install_requires=[
+
```
Comment (gmastrapas) [inline setup.py:30]: why a space here?

**Review by gmastrapas — CHANGES_REQUESTED** [review]

File: tests/unit/test___init__.py [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+    'descriptor, select_model, is_onnx, expect_error',
+    [
+        ('bert-base-cased', None, False, None),
+        ('bert-base-cased', None, True, None),
+        ('openai/clip-vit-base-patch16', 'clip-text', False, None),
+        ('openai/clip-vit-base-patch16', 'clip-vision', False, None),
+        ('openai/clip-vit-base-patch16', None, False, SelectModelRequired),
+        ('MADE UP MODEL', None, False, ValueError),
+    ],
+)
```
[the remark this answers is not in the document]
Comment (LMMilliken) [reply tests/unit/test___init__.py:18]: The more you know!

File: finetuner/__init__.py (hunk 8 above)
Comment (LMMilliken) [reply finetuner/__init__.py:317]: which argument, logging_level?

File: setup.py (hunk 11 above)
Comment (LMMilliken) [reply setup.py:30]: This must have been a mistake when merging with main, ill fix that

File: finetuner/__init__.py [hunk 13]
```diff
@@ -308,6 +308,63 @@ def get_token() -> str:
...
+    select_model: Optional[str] = None,
+    device: Optional[str] = None,
+    is_onnx: bool = False,
+):
+    """
+    Builds a pre-trained model from a given descriptor.
+
+    :param name: Refers to a pre-trained model, see
+        https://finetuner.jina.ai/walkthrough/choose-backbone/  or use the
+        :meth:`finetuner.describe_models()` functionfor a list of all
```
Comment (gmastrapas) [inline finetuner/__init__.py:324]: `function for` instead of `functionfor`

**Review by gmastrapas — COMMENTED** [review]
One minor thing

**Review by LMMilliken — COMMENTED** [review]
chore: fix typo in doc string

File: finetuner/__init__.py [hunk 14]
```diff
@@ -308,6 +308,63 @@ def get_token() -> str:
...
 
 
+def build_model(
+    name: str,
+    model_options: Optional[Dict[str, Any]] = None,
+    batch_size: int = 32,
+    select_model: Optional[str] = None,
+    device: Optional[str] = None,
+    is_onnx: bool = False,
+):
```
Comment (guenthermi) [inline finetuner/__init__.py:318]: Here is an annotation of the return type missing

**Review by gmastrapas — APPROVED** [review]
LGTM!

File: tests/unit/test___init__.py [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+        with pytest.raises(expect_error):
+            model = finetuner.build_model(
+                name=descriptor,
+                select_model=select_model,
+                is_onnx=is_onnx,
+            )
+    else:
+        model = finetuner.build_model(
+            name=descriptor, select_model=select_model, is_onnx=is_onnx
+        )
```
Comment (guenthermi) [inline tests/unit/test___init__.py:31]: if build_model returns nothing it would actually pass. Testing something with the model would be better. Maybe you can just check if you can pass an input and the result is a vector of the expected shape. If this is too much, you could at least check if the type of `model` is correct and not None

**Review by guenthermi — COMMENTED** [review]
I added too comments, rather minor

**Review by guenthermi — APPROVED** [review]
LGTM, but I think documentation for it is still missing, but you can do this in a separate PR.
